cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,804 @@
+    }
+
+    /**
+     * @param vertx The Vertx instance
+     * @param pfa Platform features availability properties
+     * @param supplier Supplies the operators for different resources
+     * @param ccHost Optional host address for the Cruise Control REST API. If this is not supplied then Cruise Control
+     *             service address will be used. This parameter is intended for use in testing.
+     */
+    public KafkaClusterRebalanceAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa,
+                                                 ResourceOperatorSupplier supplier, String ccHost) {
+        this(vertx, pfa, supplier, v -> new CruiseControlApiImpl(vertx), ccHost);
+    }
+
+    public KafkaClusterRebalanceAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa,
+                                                 ResourceOperatorSupplier supplier,
+                                                 Function<Vertx, CruiseControlApi> cruiseControlClientProvider, String ccHost) {
+        super(vertx, KafkaClusterRebalance.RESOURCE_KIND, supplier.kafkaClusterRebalanceOperator, supplier.metricsProvider);
+        this.pfa = pfa;
+        this.clusterRebalanceOperator = supplier.kafkaClusterRebalanceOperator;
+        this.kafkaOperator = supplier.kafkaOperator;
+        this.cruiseControlClientProvider = cruiseControlClientProvider;
+        this.ccHost = ccHost;
+    }
+
+    /**
+     * Create a watch on {@code KafkaClusterRebalance} in the given {@code watchNamespaceOrWildcard}.
+     *
+     * @param watchNamespaceOrWildcard The namespace to watch, or "*" to watch all namespaces.
+     * @return A future which completes when the watch has been set up.
+     */
+    public Future<Void> createClusterRebalanceWatch(String watchNamespaceOrWildcard) {
+
+        return Util.async(this.vertx, () -> {
+            this.clusterRebalanceOperator.watch(watchNamespaceOrWildcard, new Watcher<KafkaClusterRebalance>() {
+                @Override
+                public void eventReceived(Action action, KafkaClusterRebalance kafkaClusterRebalance) {
+                    log.debug("EventReceived {} on {} with status [{}] and {}={}", action, kafkaClusterRebalance.getMetadata().getName(),
+                            kafkaClusterRebalance.getStatus() != null ? kafkaClusterRebalance.getStatus().getConditions().get(0).getType() : null,
+                            ANNO_STRIMZI_IO_REBALANCE, rebalanceAnnotation(kafkaClusterRebalance));
+                    String clusterName = kafkaClusterRebalance.getMetadata().getLabels() == null ? null : kafkaClusterRebalance.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
+                    String clusterNamespace = kafkaClusterRebalance.getMetadata().getNamespace();
+                    KafkaClusterRebalanceStatus desiredStatus = new KafkaClusterRebalanceStatus();
+                    if (clusterName != null) {
```

Comment:
It looks like you execute this logic here only when event is triggered. Maybe I missed something, but this looks wrong. You have your own reconcileAll loop and you need to test this everytime in the loop, or?

Reply:
good catch!